### DISCUSSION TITLE
Introspection on vsc.randobj

### DISCUSSION THREAD

**Discussion by msmftc** [fvutils/pyvsc#135] [Q&A]
@mballance,

In my application I need some introspection on classes that are decorated with `@vsc.randoj`.  Given a reference to such a class, I need to determine the module that the class was defined in.  The class property `__module__` typically has this information, but decorating the class changes `__module__` to `vsc.rand_obj`

Based on advice I found on the internet, I tried adding a `@functools.wraps` decorator inside the function `vsc.randobj()`, but that did not help.  The only solution that I've found is to add a new line of code to `vsc.randobj()` that stores the original module name in a new property, as shown here:

```python
def randobj(*args, **kwargs):
    if len(args) == 1 and len(kwargs) == 0 and callable(args[0]):
        # Called without arguments
        obj = _randobj({})
        args[0].original_module = args[0].__module__  # NEW LINE OF CODE
        return obj(args[0])
    else:
        obj = _randobj(kwargs)
        return obj
```

This lets me read the original module name, as shown in the example below, but I'd rather not use a hacked version of PyVSC.  Do you know how I can read the original module using the current PyVSC code?  If not, then can you think of a simple PyVSC change that would permit introspection of decorated classes?

In this example, the original module name is `__main__` for both `HelloClass` and `Selector`:
```python
import vsc

class HelloClass:
    string = "Hello"

@vsc.randobj
class Selector:
    def __init__(self):
        self.a = vsc.rand_uint16_t()


print(f"HelloClass Module =          {HelloClass.__module__}")
print(f"Selector Module =            {Selector.__module__}")
print(f"Selector original_module =   {Selector.original_module}")
```
```
HelloClass Module =          __main__
Selector Module =            vsc.rand_obj
Selector original_module =   __main__
```

**Comment by mballance** [comment]
@msmftc, thanks for raising this. In general, the goal of `vsc.randobj` is to be transparent. Certainly, it adds some functions and behavior to a user class, but it is intended to (largely) leave the user's view of the class undisturbed. The example you show about highlights a situation where this isn't the case.
Let me give this a bit more thought, but my initial reaction is that the `vsc.randobj` decorator should propagate attributes such as `__module__` back to the new class defined by `vsc.randobj`. My biggest concern is what attributes, beyond `__module__`, will be overlooked and need to be added incrementally...

**Comment by mballance** [comment]
Note: this has been addressed in the 0.8.1 release
